File: gs_knizia/bethe_ansatz.cpp
/* This program is released under a "I hope no one has to deal with this
 * ever again"-license. You may do with it as you wish, including
 * distributing it in source- or binary form, as long as you do not blame
 * me if anything goes wrong.
 * 
 * If you find the program useful, I would appreciate a citation to
 * 
 *    Knizia, Chan - Density Matrix Embedding: A Simple Alternative
 *       to Dynamical Mean-Field Theory
 *    http://dx.doi.org/10.1103/PhysRevLett.109.186404
 * 
 * in published work using the program. That is the paper for which it
 * was written. In any case you should cite
 * 
 *    Shiba - Magnetic Susceptibility at Zero Temperature for the
 *       One-Dimensional Hubbard Model
 *    http://dx.doi.org/10.1103/PhysRevB.6.930
 * 
 * which describes the algorithm which was employed in this program.
 * -- Gerald Knizia, 2014-06-03
 */

// calculates 1D Hubbard model E(U,f) where f is the filling; based on
// [1] Shiba - Magnetic susceptibility at zero temperture for the
//     one-dimensional Hubbard model [Phys Rev B 6 930 (1972)]
//
// License: Use as you see fit, but don't blame me if something goes
//          wrong!
//
//                                   -- Gerald Knizia, 2011
//
// Compile:
//    c++ -O3 -DNDEBUG bethe_ansatz.cpp -o bethe_ansatz
// Run:
//    Meant to be run by bethe_ansatz.py.
//    This program just takes two arguments (U and Q) and calculates
//    the energy and the particle number resulting from this. Other properties
//    can be obtained as derivatives from those.
//
// See README.txt

#include <iostream>
#include <cmath>
#include <cstring>
#include <stdexcept>
#include <cstdlib> // atof/atoi
#include <vector>
#include <boost/format.hpp>

#ifndef M_PI
   #define M_PI 3.14159265358979323846
#endif

using std::cout;
using std::endl;
using boost::format;
using boost::str;

typedef double
   scalar;
typedef unsigned int
   uint;

#define USE_ALTERNATING_SUM_ACCEL
#if not defined(USE_ALTERNATING_SUM_ACCEL)
   // Very slow. This branch just for testing.

   // [0] eq. 2.7.
   static scalar R(scalar x)
   {
      scalar
         r = 0,
         sign = 1.0;  // start at + for n == 1.
      for (scalar n = 1; n < 2000000 ; ++ n) {
         scalar d = 2*n/(x*x + 4*n*n);
         r += sign * d;
         sign = -sign;
         if ((((long)n) & 0xfff) == 1)
            cout << (format("   n = %i   r = %.8f   d = %f") % n % r % d) << endl;
         if (std::abs(d/r) < 1e-16)
            break;
      }
      return r/M_PI;
   }
#else
   // Convergence acceleration of alternating series
   // Henri Cohen, Fernando Rodriguez Villegas, and Don Zagier
   // Experiment. Math. Volume 9, Issue 1 (2000), 3-12.
   typedef scalar (*sum_fn_t)(uint k, double x);

   inline scalar r_k(uint n, scalar x){
      n += 1; // summation starts at n=0 here with positive sign, that's why we shift from (2.7) by one.
      return 2*n/(x*x + 4*n*n);
   }

   static scalar alternating_sum(uint n, sum_fn_t an, scalar param)
   {
      scalar
         b = -1, c, d, s = 0;
      d = std::pow((3. + std::sqrt(8.)), n);
      d = (d+1/d)/2;
      c = -d;
      for (scalar k = 0; k < n; ++ k) {
         c = b - c;
         s = s + c * an(k, param);
         b *= (k + n)*(k - n)/((k+.5)*(k+1));
      }
      return s/d;
   }

   static scalar R(scalar x)
   {
      scalar e2 = alternating_sum(22, r_k, x);
   #ifdef _DEBUG
      scalar e1 = alternating_sum(20, r_k, x),
      if (abs(e1-e2) > 1e-15) throw std::runtime_error("N too small in R(x)!");
   #endif // _DEBUG
      return e2/M_PI;
   }
#endif // USE_ALTERNATING_SUM_ACCEL


uint
   // we use an equidistant grid. That is fine in this case, because the
   // function to integrate falls of quickly at the boundaries and is also
   // otherwise very well-behaved. Higher order integration rules are
   // unnecessary, and the 200pt grid is also plenty.
   nGrid = 200;

// grid positions x(k)
inline scalar x(uint k, scalar Q) {
   return Q*(2*(signed)k + 1 - (signed)nGrid)/static_cast<scalar>(nGrid);
}

// integration weight dx(k)
inline scalar dx(uint k, scalar Q) {
   return (2.*Q)/nGrid;
}

// inline scalar dx(uint k, scalar Q) {
//    scalar f0 = (1./3.)*(2.*Q)/nGrid;
//    assert(nGrid % 2 == 1 );
//    if ( k == 0 || k == nGrid -1 ) return f0;
//    if ( k % 2 == 1 ) return 4.*f0;
//    else return 2.*f0;
// };
// ^- hm... not very convincing.


static scalar eval_rho_rhs_for_Q(scalar &EperSite, scalar &NbyNa, scalar *rho_out, scalar *rho_in, scalar Q, scalar u, std::vector<scalar> &tmp)
{
   // evaluates one iteration of [0] eq. 2.8 for a given rho, on the
   // interval [-Q,Q] with nGrid grid points.
   scalar
      invu = 1./u,
      max_dev = 0; // maximal change in rho.
   EperSite = 0; // E/Na
   NbyNa = 0;
   tmp.resize(nGrid*2);
   scalar
      *cosxi = &tmp[0],
      *sinxi = &tmp[nGrid];
   for (uint i = 0; i < nGrid; ++ i) {
      scalar xi = x(i,Q);
      cosxi[i] = std::cos(xi);
      sinxi[i] = std::sin(xi);
   }

   for (uint i = 0; i < nGrid; ++ i) {
      scalar fint = 0;
      for (uint k = 0; k < nGrid; ++ k) {
         fint += dx(k,Q) * rho_in[k] * R(4.*invu*(sinxi[i] - sinxi[k]));
      }
      fint *= 8.*M_PI*invu;
      max_dev = std::max(std::abs(rho_out[i] - rho_in[i]), max_dev);
      rho_out[i] = (1.0/(2.*M_PI)) * (1 + cosxi[i] * fint);
      EperSite += -2 * dx(i,Q) * cosxi[i] * rho_out[i];
      NbyNa += dx(i,Q) * rho_out[i];
   }
   return max_dev;
}


struct result_entry_t {
   scalar u;
   scalar filling;
   scalar energy_per_site;
   scalar Q;
};

static void solve_for_rho(result_entry_t *r, scalar *rho, scalar Q, scalar u, int print = 1)
{
   double dabl = 0.;
   for (uint i = 0; i < nGrid; ++ i)
      dabl += dx(i,Q);
   if (print >= 2) {
      cout << format("sum[dx] = %.8f  2*Q=%.8f") % dabl % (2*Q) << endl;
      cout << format("x[0] = %.8f") % x(0,Q) << endl;
      cout << format("x[nGrid-1] = %.8f") % x(nGrid-1,Q) << endl;
   }
   r->Q = Q;
   r->u = u;

   bool
      converged = false;
   std::vector<scalar>
      dummy(nGrid, 0.0),
      tmp;
   scalar
      *pin = &dummy[0], *pout = rho, last_e = 1e99;
   for (uint it = 0; it < 1000; ++ it) {
      double max_dev_rho = eval_rho_rhs_for_Q(r->energy_per_site, r->filling, pout, pin, Q, u, tmp);
      double delta_e = last_e - r->energy_per_site;
      if (print >= 1) {
         cout << format("it = %-4i  E = %10.5f  N/Na = %10.5f  max_dev_rho = %7.2e  delta_e = %7.2e  rho = [")
            % it % r->energy_per_site % r->filling % max_dev_rho % delta_e;
         uint nPrint = 7;
         for ( uint i = 0; i < nPrint; ++ i )
            cout << format(" %10.5f") % pout[(i*nGrid)/(nPrint-1)];
         cout << "]" << endl;
      }
      std::swap(pin, pout);
      if (it > 2 && max_dev_rho < 1e-15 && std::abs(delta_e) < 1e-15) {
         converged = true;
         break;
      }
      last_e = r->energy_per_site;
   }
   if (!converged)
      throw std::runtime_error(str(format("sorry, bethe_ansatz.cpp failed to converge! Q=%.8f u=%8f.") % Q % u));
   if (rho != pin)
      std::memcpy(rho, pin, sizeof(rho[0])*nGrid);
}

int main(int argc, char **argv)
{
   // note: t is fixed at 1.0.

   if (argc != 3)
      throw std::runtime_error("usage: bethe <U> <Q>    with U, Q floating point numbers. Q \\in [0,1]. U < 0.05 should not be used.");
   scalar u = atof(argv[1]);
   scalar q = atof(argv[2]);

   // for small U we need more grid points---need to accurately
   // sample R(4*sin(x)/u) for x = -pi..pi.
   if (u < 1.) {
      nGrid *= 1./u;
      nGrid += (nGrid % 2);
      std::cout << format("WARNING: Increased number of grid points to %i due to small U. May get slow!") % nGrid << std::endl;
   }

   std::vector<scalar>
      rho(nGrid, 0.0);
   result_entry_t r;
   solve_for_rho(&r, &rho[0], q * M_PI, u, 0);
   std::cout << format("E= %.15f  n= %.15f") % r.energy_per_site % r.filling << std::endl;
   return 0;
}


// Note on the source formatting:
//    cgk: sometimes I try new coding styles on small projects, to see
//    if I like them. Turns out that I do not like this one at all.


//
// (note: see also Cole III - ``Excitation spectrum of the one-dimensional
//  Hubbard model'' for some more information.
//  On excited states: can't we just take higher eigenvectors of the
//  rho self-consistency equation? Since rho has a maximum at k=0, it seems
//  that we should just solve for higher distributions. But what about the
//  E = int cos(k) rho(k)  instead of E = int rho[k]?.. we need orthogonal
//  wave functions, not density distributions. However, Cole's approach
//  does not apply to the half filled case (Woynarovich, J Phys C 15 85 (1982)))


// kate: indent-width 3

File: gs_knizia/bethe_ansatz.py
# This program is released under a "I hope no one has to deal with this
# ever again"-license. You may do with it as you wish, including
# distributing it in source- or binary form, as long as you do not blame
# me if anything goes wrong.
# 
# If you find the program useful, I would appreciate a citation to
# 
#    Knizia, Chan - Density Matrix Embedding: A Simple Alternative
#       to Dynamical Mean-Field Theory
#    http://dx.doi.org/10.1103/PhysRevLett.109.186404
# 
# in published work using the program. That is the paper for which it
# was written. In any case you should cite
# 
#    Shiba - Magnetic Susceptibility at Zero Temperature for the
#       One-Dimensional Hubbard Model
#    http://dx.doi.org/10.1103/PhysRevB.6.930
# 
# which describes the algorithm which was employed in this program.
# -- Gerald Knizia, 2014-06-03

from __future__ import print_function
import numpy as np
import scipy.optimize as optimize
from subprocess import getoutput
from math import pi

CALL_DIR = './'

def CalcBetheEnergy_UandQ(U, Q_):
   """invoke the C++ program for given parameters U, Q. Return energy for
   those parameters. Q is given in units of radians and must be in range
   of (0...2) (corresponding to just-above zero to just below-full filling)
   """
   if Q_ > 1:
      # "code" for N > 1. Transform to the equivalent system with N < 1.
      # (iterative algorithm works only for N <= 1)
      E, N = CalcBetheEnergy_UandQ(U, 2. - Q_)
      return 2*((1.-N/2)-.5)*U + E, 2. - N

   # clamp to valid range. Q=0 is empty, Q=1.0 is half-filling.
   Q = np.clip(Q_, 0.0, 1.0)
   Text = getoutput(CALL_DIR+"bethe_ansatz %.15f %.15f" % (U, Q))
   Lines = Text.splitlines()[-1]
   ls = Lines.split()
   E = float(ls[1])
   N = float(ls[3])
   return E, N

def FindQForN(U, TargetN):
   """Find the Q(U,n) parameter which will result in a filling fraction
   of <n> = TargetN for the given U. Note: This is done by numerical root
   search."""
   assert(0. <= TargetN <= 2.)
   if TargetN > 1:
      return 2. - FindQForN(U, 2.-TargetN)
   if TargetN == 1.0:
      return 1.  # no need to search for that.
   Tol = 1e-10
   def Err(Q):
      E, N = CalcBetheEnergy_UandQ(U, Q)
      return (N - TargetN)**2
   try:
      Q = optimize.brent(Err, brack = (0.0, 0.95, 1.0), tol=Tol, maxiter=1000)
   except ValueError:
      Q = optimize.fminbound(Err, 0.0, 1.0, xtol=Tol, maxfun=1000)
   E, N = CalcBetheEnergy_UandQ(U, Q)
   assert(abs(N - TargetN) <= 1e-5)
   return Q

def CalcBetheEnergy_UandN(U, N):
   """as CalcBetheEnergy_UandQ(U,Q), but use U and <n> as input. This requires first
   solving for Q such that <n> = n(U,Q) and is consequently much slower
   than CalcBetheEnergy_UandQ."""
   return CalcBetheEnergy_UandQ(U, FindQForN(U,N))

def GetLeftMuAtHalfFilling(U):
   """Compute derivative d/d[n] E(U,n) for limit n -> 1.0 - 0
   (i.e., on left side of n=1)"""
   Ql = FindQForN(U, 1.0 - 1e-7)
   Em, Nm = CalcBetheEnergy_UandQ(U, Ql)
   E0, N0 = CalcBetheEnergy_UandQ(U, 1.0)
   return (E0-Em)/(N0-Nm)

def FindGaps():
   """Compute only the one-particle gaps at half filling."""
   TargetUs = [4.]
   for U in TargetUs:
      MuLeft = GetLeftMuAtHalfFilling(U)
      MuHalf = U/2
      Gap = 2 * (MuHalf - MuLeft)
      print("U = {:.4f}  MuLeft = {:.6f}  MuHalf = {:.6f}  Gap = {:.6f}".format(U, MuLeft, MuHalf, Gap))

def MakeSpinCorrelationCurves():
   """Make table of spin correlation functions."""
   ResultLines = []

   TargetUs = []
   TargetUs = [4.,2.,1.,0.5,0.25,0.125]
   #Ns = [.25,1./3.,.5,2./3.,1.0]
   #Ns = [1./3.,2./3., .25,.5,.75, 1.]
   Ns = [1./3.,2./3.,1.]

   def SzSz_for_NaNb(N, NaNb):
      """find <Sz(i) Sz(i)> correlation function for given <nA(i) nB(i)> correlation
      function and filling <n>."""
      SzSz = 3*(N - 2*NaNb)/4
      # ^- PRB 37 7541 (1988) eq 33 (in vector form in that paper)
      #    cites Shiba PRB 6 930, eq 2.12

      # FIXME: hm...  Shibas' correlation functions for <\vec S^2> go from 3/8 to 3/4;
      # mine for <Sz.Sz> from .5 to 1.0. The numbers appear to agree perfectly,
      # but I am not sure why I should divide by (3/4)!
      # That seems to be some normalization factor for S^2. SzSz = N - 2*NaNb makes
      # sense to me (w/o the 3/4.). I just accept it here...
      SzSz /= (3./4.)
      return SzSz

   if (tuple(Ns) == (1./3., 2./3., 1.)):
      ResultLines.append("       0.000   0.333333   0.027778 {:10.6f} {:10.6f}  * note: manual. <na.nb> = 1/(9.*4)".format(1/(9.*4), SzSz_for_NaNb(1/3., 1/(9.*4))))
      ResultLines.append("       0.000   0.666667   0.111111 {:10.6f} {:10.6f}  * note: manual. <na.nb> = 1/9.".format(1/9., SzSz_for_NaNb(2/3.,1/9.)))
      ResultLines.append("       0.000   1.000000   0.250000 {:10.6f} {:10.6f}  * note: manual. <na.nb> = 1/4.".format(1/4., SzSz_for_NaNb(1., 1/4.)))
      # ^- exact solutions for some boundary cases without Us.

   for N in Ns:
      for U in TargetUs:
         if U == 0.:
            continue

         # <nalpha nbeta> = d/dU E(U,n)
         h = 0.001
         Ep,Np = CalcBetheEnergy_UandN(U+h, N)
         Em,Nm = CalcBetheEnergy_UandN(U-h, N)
         E0,N0 = CalcBetheEnergy_UandN(U,N)
         NaNb = (Ep-Em)/(2*h)
         SzSz = SzSz_for_NaNb(N, NaNb)
         print("U = {:10.3f}  <n> = {:10.4f}  <na.nb> = {:10.4f}  <Sz.Sz> = {:10.4f}   {:.2e} {:.6f} {:.6f}".format(U, N, NaNb, SzSz, Ep-Em, Np, Nm))
         ResultLines.append("  {:10.3f} {:10.6f} {:10.6f} {:10.6f} {:10.6f}".format(U, N, E0, NaNb, SzSz))
   print()
   print("  {:>10} {:>10} {:>10} {:>10} {:>10}".format("U  ", "<n>  ", "E  ", "<na.nb>", "<Sz.Sz>"))
   print("   " + 5 * "-----------")
   ResultLines.sort()
   print("\n".join(ResultLines))

def MakeEnergyCurvesAtHalfFilling():
   print("     U         Filling       E/Site         Q/pi")
   print("---------------------------------------------------")
   for U in np.arange(0.25,128.01,0.25):
      for Q in [1.]:
         E, N = CalcBetheEnergy_UandQ(U, Q)
         print("{:8.3f} {:13.6f} {:13.6f} {:13.6f}".format(U, .5*N, E, Q))
      #print

def MakeCurvesForQ():
   """Compute E(U,n) curves, with N given implicitly as a function of Q. (i.e.,
   the curves will be a representation of the full E(U,n) range, but they will have
   different N coordinates for different U).
   Note that the Energy is supposed to have a spike and mu a gap """
   Cap = "   {:^7s} {:^13s} {:^13s} {:^13s} {:^13s}".format("U", "Filling", "E/Site", "Chem.Pot", "Q/pi")
   print(Cap + "\n" + len(Cap)*"-")
   #for U in np.arange(0.25,32.01,0.25):
   for U in [4.]:
      for Q in np.linspace(0.0, 2.0, 161):
      #for Q in np.arange(0.98,1.02+1e-8,0.001):
         dQ = 1e-3
         if Q <= dQ or (Q >= 1.0 - dQ and Q <= 1.0 + dQ):
            continue # can't make numerical derivatives there.
         Ep, Np = CalcBetheEnergy_UandQ(U, Q+dQ)
         Ec, Nc = CalcBetheEnergy_UandQ(U, Q)
         Em, Nm = CalcBetheEnergy_UandQ(U, Q-dQ)
         Mu = (Ep-Em)/(Np-Nm)
         print("{:8.3f} {:13.6f} {:13.6f} {:13.6f} {:13.6f}".format(U, .5*Nc, Ec, Mu, Q))
      print()

def MakeRefDataAtHalfFilling():
   """Compute various information relevant at and around half filling."""
   print("     U        Filling       E/Site        HL-Gap        Chem.Pot      <na.nb>")
   print("------------------------------------------------------------------------------")
   L2 = []
   for U in np.arange(1.,16.+1e-8,1.):
      Q = 1.
      dQ = 1e-3
      Eq, Nq = CalcBetheEnergy_UandQ(U, Q+2*dQ)
      Ep, Np = CalcBetheEnergy_UandQ(U, Q+dQ)
      Ec, Nc = CalcBetheEnergy_UandQ(U, Q)
      Em, Nm = CalcBetheEnergy_UandQ(U, Q-dQ)
      En, Nn = CalcBetheEnergy_UandQ(U, Q-2*dQ)

      E = Ec
      MuN = (Ep-Eq)/(Np-Nq)
      MuP = (En-Em)/(Nn-Nm)
      Gap = MuN - MuP
      Mu = (Ep-Em)/(Np-Nm)

      h = 0.001
      Fp,Mp = CalcBetheEnergy_UandQ(U+h, Q)
      Fm,Mm = CalcBetheEnergy_UandQ(U-h, Q)
      F0,M0 = CalcBetheEnergy_UandQ(U,Q)
      D = (Fp-Fm)/(2*h)

      print("{:8.3f}  {:12.6f}                              {:12.6f}".format(U, Nn/2., MuP))
      print("{:8.3f}  {:12.6f}  {:12.6f}  {:12.6f}  {:12.6f}  {:12.6f}".format(U, Nc/2., Ec, Gap, Mu, D))


# choose the computation to run.
Fn = [FindGaps, MakeSpinCorrelationCurves, MakeEnergyCurvesAtHalfFilling, MakeCurvesForQ, MakeRefDataAtHalfFilling][-1]
Fn()
